// antlr4-c3/out/src/SymbolTable.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SymbolTable = exports.ArrayType = exports.InterfaceSymbol = exports.ClassSymbol = exports.FieldSymbol = exports.MethodSymbol = exports.MethodFlags = exports.RoutineSymbol = exports.ParameterSymbol = exports.LiteralSymbol = exports.VariableSymbol = exports.BlockSymbol = exports.NamespaceSymbol = exports.ScopedSymbol = exports.TypeAlias = exports.TypedSymbol = exports.Symbol = exports.FundamentalType = exports.ReferenceKind = exports.TypeKind = exports.Modifier = exports.MemberVisibility = exports.DuplicateSymbolError = void 0;
class DuplicateSymbolError extends Error {
}
exports.DuplicateSymbolError = DuplicateSymbolError;
var MemberVisibility;
(function (MemberVisibility) {
    MemberVisibility[MemberVisibility["Unknown"] = 0] = "Unknown";
    MemberVisibility[MemberVisibility["Open"] = 1] = "Open";
    MemberVisibility[MemberVisibility["Public"] = 2] = "Public";
    MemberVisibility[MemberVisibility["Protected"] = 3] = "Protected";
    MemberVisibility[MemberVisibility["Private"] = 4] = "Private";
    MemberVisibility[MemberVisibility["FilePrivate"] = 5] = "FilePrivate";
    MemberVisibility[MemberVisibility["Library"] = 6] = "Library";
})(MemberVisibility = exports.MemberVisibility || (exports.MemberVisibility = {}));
var Modifier;
(function (Modifier) {
    Modifier[Modifier["Static"] = 0] = "Static";
    Modifier[Modifier["Final"] = 1] = "Final";
    Modifier[Modifier["Sealed"] = 2] = "Sealed";
    Modifier[Modifier["Abstract"] = 3] = "Abstract";
    Modifier[Modifier["Deprecated"] = 4] = "Deprecated";
    Modifier[Modifier["Virtual"] = 5] = "Virtual";
    Modifier[Modifier["Const"] = 6] = "Const";
    Modifier[Modifier["Overwritten"] = 7] = "Overwritten";
})(Modifier = exports.Modifier || (exports.Modifier = {}));
var TypeKind;
(function (TypeKind) {
    TypeKind[TypeKind["Unknown"] = 0] = "Unknown";
    TypeKind[TypeKind["Integer"] = 1] = "Integer";
    TypeKind[TypeKind["Float"] = 2] = "Float";
    TypeKind[TypeKind["Number"] = 3] = "Number";
    TypeKind[TypeKind["String"] = 4] = "String";
    TypeKind[TypeKind["Char"] = 5] = "Char";
    TypeKind[TypeKind["Boolean"] = 6] = "Boolean";
    TypeKind[TypeKind["Class"] = 7] = "Class";
    TypeKind[TypeKind["Interface"] = 8] = "Interface";
    TypeKind[TypeKind["Array"] = 9] = "Array";
    TypeKind[TypeKind["Map"] = 10] = "Map";
    TypeKind[TypeKind["Enum"] = 11] = "Enum";
    TypeKind[TypeKind["Alias"] = 12] = "Alias";
})(TypeKind = exports.TypeKind || (exports.TypeKind = {}));
var ReferenceKind;
(function (ReferenceKind) {
    ReferenceKind[ReferenceKind["Irrelevant"] = 0] = "Irrelevant";
    ReferenceKind[ReferenceKind["Pointer"] = 1] = "Pointer";
    ReferenceKind[ReferenceKind["Reference"] = 2] = "Reference";
    ReferenceKind[ReferenceKind["Instance"] = 3] = "Instance";
})(ReferenceKind = exports.ReferenceKind || (exports.ReferenceKind = {}));
class FundamentalType {
    constructor(name, typeKind = TypeKind.Unknown, referenceKind = ReferenceKind.Irrelevant) {
        this.name = name;
        this.typeKind = typeKind;
        this.referenceKind = referenceKind;
    }
    get baseTypes() {
        return [];
    }
    get kind() {
        return this.typeKind;
    }
    get reference() {
        return this.referenceKind;
    }
}
FundamentalType.integerType = new FundamentalType("int", TypeKind.Integer, ReferenceKind.Instance);
FundamentalType.floatType = new FundamentalType("float", TypeKind.Float, ReferenceKind.Instance);
FundamentalType.stringType = new FundamentalType("string", TypeKind.String, ReferenceKind.Instance);
FundamentalType.boolType = new FundamentalType("bool", TypeKind.Boolean, ReferenceKind.Instance);
exports.FundamentalType = FundamentalType;
class Symbol {
    constructor(name = "") {
        this.name = "";
        this.modifiers = new Set();
        this.visibility = MemberVisibility.Unknown;
        this.name = name;
    }
    get parent() {
        return this.theParent;
    }
    get firstSibling() {
        if (this.theParent instanceof ScopedSymbol) {
            return this.theParent.firstChild;
        }
        return this;
    }
    get previousSibling() {
        if (!(this.theParent instanceof ScopedSymbol)) {
            return this;
        }
        return this.theParent.previousSiblingOf(this);
    }
    get nextSibling() {
        if (!(this.theParent instanceof ScopedSymbol)) {
            return this;
        }
        return this.theParent.nextSiblingOf(this);
    }
    get lastSibling() {
        if (this.theParent instanceof ScopedSymbol) {
            return this.theParent.lastChild;
        }
        return this;
    }
    get next() {
        if (this.theParent instanceof ScopedSymbol) {
            return this.theParent.nextOf(this);
        }
        return undefined;
    }
    get root() {
        let run = this.theParent;
        while (run) {
            if (!run.theParent || (run.theParent instanceof SymbolTable)) {
                return run;
            }
            run = run.theParent;
        }
        return run;
    }
    get symbolTable() {
        if (this instanceof SymbolTable) {
            return this;
        }
        let run = this.theParent;
        while (run) {
            if (run instanceof SymbolTable) {
                return run;
            }
            run = run.theParent;
        }
        return undefined;
    }
    get symbolPath() {
        const result = [];
        let run = this;
        while (run) {
            result.push(run);
            if (!run.theParent) {
                break;
            }
            run = run.theParent;
        }
        return result;
    }
    setParent(parent) {
        this.theParent = parent;
    }
    removeFromParent() {
        if (this.theParent instanceof ScopedSymbol) {
            this.theParent.removeSymbol(this);
            this.theParent = undefined;
        }
    }
    resolve(name, localOnly = false) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.theParent instanceof ScopedSymbol) {
                return this.theParent.resolve(name, localOnly);
            }
            return Promise.resolve(undefined);
        });
    }
    resolveSync(name, localOnly = false) {
        if (this.theParent instanceof ScopedSymbol) {
            return this.theParent.resolveSync(name, localOnly);
        }
        return undefined;
    }
    getParentOfType(t) {
        let run = this.theParent;
        while (run) {
            if (run instanceof t) {
                return run;
            }
            run = run.theParent;
        }
        return undefined;
    }
    qualifiedName(separator = ".", full = false, includeAnonymous = false) {
        if (!includeAnonymous && this.name.length === 0) {
            return "";
        }
        let result = this.name.length === 0 ? "<anonymous>" : this.name;
        let run = this.theParent;
        while (run) {
            if (includeAnonymous || run.name.length > 0) {
                result = (run.name.length === 0 ? "<anonymous>" : run.name) + separator + result;
            }
            if (!full || !run.theParent) {
                break;
            }
            run = run.theParent;
        }
        return result;
    }
}
exports.Symbol = Symbol;
class TypedSymbol extends Symbol {
    constructor(name, type) {
        super(name);
        this.type = type;
    }
}
exports.TypedSymbol = TypedSymbol;
class TypeAlias extends Symbol {
    constructor(name, target) {
        super(name);
        this.targetType = target;
    }
    get baseTypes() {
        return [this.targetType];
    }
    get kind() {
        return TypeKind.Alias;
    }
    get reference() {
        return ReferenceKind.Irrelevant;
    }
}
exports.TypeAlias = TypeAlias;
class ScopedSymbol extends Symbol {
    constructor(name = "") {
        super(name);
        this.childSymbols = [];
    }
    get directScopes() {
        return this.getSymbolsOfType(ScopedSymbol);
    }
    get children() {
        return this.childSymbols;
    }
    get firstChild() {
        if (this.children.length > 0) {
            return this.children[0];
        }
        return undefined;
    }
    get lastChild() {
        if (this.children.length > 0) {
            return this.children[this.children.length - 1];
        }
        return undefined;
    }
    clear() {
        this.childSymbols = [];
    }
    addSymbol(symbol) {
        symbol.removeFromParent();
        const symbolTable = this.symbolTable;
        if (!symbolTable || !symbolTable.options.allowDuplicateSymbols) {
            this.children.forEach((child) => {
                if (child === symbol || (symbol.name.length > 0 && child.name === symbol.name)) {
                    let name = symbol.name;
                    if (name.length === 0) {
                        name = "<anonymous>";
                    }
                    throw new DuplicateSymbolError("Attempt to add duplicate symbol '" + name + "'");
                }
            });
        }
        this.children.push(symbol);
        symbol.setParent(this);
    }
    removeSymbol(symbol) {
        const index = this.children.indexOf(symbol);
        if (index > -1) {
            this.children.splice(index, 1);
            symbol.setParent(undefined);
        }
    }
    getNestedSymbolsOfType(t) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = [];
            const childPromises = [];
            this.children.forEach((child) => {
                if (child instanceof t) {
                    result.push(child);
                }
                if (child instanceof ScopedSymbol) {
                    childPromises.push(child.getNestedSymbolsOfType(t));
                }
            });
            const childSymbols = yield Promise.all(childPromises);
            childSymbols.forEach((entry) => {
                result.push(...entry);
            });
            return result;
        });
    }
    getNestedSymbolsOfTypeSync(t) {
        const result = [];
        this.children.forEach((child) => {
            if (child instanceof t) {
                result.push(child);
            }
            if (child instanceof ScopedSymbol) {
                result.push(...child.getNestedSymbolsOfTypeSync(t));
            }
        });
        return result;
    }
    getAllNestedSymbols(name) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = [];
            const childPromises = [];
            this.children.forEach((child) => {
                if (!name || child.name === name) {
                    result.push(child);
                }
                if (child instanceof ScopedSymbol) {
                    childPromises.push(child.getAllNestedSymbols(name));
                }
            });
            const childSymbols = yield Promise.all(childPromises);
            childSymbols.forEach((entry) => {
                result.push(...entry);
            });
            return result;
        });
    }
    getAllNestedSymbolsSync(name) {
        const result = [];
        this.children.forEach((child) => {
            if (!name || child.name === name) {
                result.push(child);
            }
            if (child instanceof ScopedSymbol) {
                result.push(...child.getAllNestedSymbolsSync(name));
            }
        });
        return result;
    }
    getSymbolsOfType(t) {
        return new Promise((resolve) => {
            const result = [];
            this.children.forEach((child) => {
                if (child instanceof t) {
                    result.push(child);
                }
            });
            resolve(result);
        });
    }
    getAllSymbols(t, localOnly = false) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = [];
            for (const child of this.children) {
                if (child instanceof t) {
                    result.push(child);
                }
                if (child instanceof NamespaceSymbol) {
                    const childSymbols = yield child.getAllSymbols(t, true);
                    result.push(...childSymbols);
                }
            }
            if (!localOnly) {
                if (this.parent instanceof ScopedSymbol) {
                    const childSymbols = yield this.getAllSymbols(t, true);
                    result.push(...childSymbols);
                }
            }
            return result;
        });
    }
    getAllSymbolsSync(t, localOnly = false) {
        const result = [];
        for (const child of this.children) {
            if (child instanceof t) {
                result.push(child);
            }
            if (child instanceof NamespaceSymbol) {
                const childSymbols = child.getAllSymbolsSync(t, true);
                result.push(...childSymbols);
            }
        }
        if (!localOnly) {
            if (this.parent instanceof ScopedSymbol) {
                const childSymbols = this.getAllSymbolsSync(t, true);
                result.push(...childSymbols);
            }
        }
        return result;
    }
    resolve(name, localOnly = false) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                for (const child of this.children) {
                    if (child.name === name) {
                        resolve(child);
                        return;
                    }
                }
                if (!localOnly) {
                    if (this.parent instanceof ScopedSymbol) {
                        resolve(this.parent.resolve(name, false));
                        return;
                    }
                }
                resolve(undefined);
            });
        });
    }
    resolveSync(name, localOnly = false) {
        for (const child of this.children) {
            if (child.name === name) {
                return child;
            }
        }
        if (!localOnly) {
            if (this.parent instanceof ScopedSymbol) {
                return this.parent.resolveSync(name, false);
            }
        }
        return undefined;
    }
    getTypedSymbols(localOnly = true) {
        const result = [];
        for (const child of this.children) {
            if (child instanceof TypedSymbol) {
                result.push(child);
            }
        }
        if (!localOnly) {
            if (this.parent instanceof ScopedSymbol) {
                const localList = this.parent.getTypedSymbols(true);
                result.push(...localList);
            }
        }
        return result;
    }
    getTypedSymbolNames(localOnly = true) {
        const result = [];
        for (const child of this.children) {
            if (child instanceof TypedSymbol) {
                result.push(child.name);
            }
        }
        if (!localOnly) {
            if (this.parent instanceof ScopedSymbol) {
                const localList = (this.parent).getTypedSymbolNames(true);
                result.push(...localList);
            }
        }
        return result;
    }
    symbolFromPath(path, separator = ".") {
        const elements = path.split(separator);
        let index = 0;
        if (elements[0] === this.name || elements[0].length === 0) {
            ++index;
        }
        let result = this;
        while (index < elements.length) {
            if (!(result instanceof ScopedSymbol)) {
                return undefined;
            }
            const child = result.children.find((candidate) => candidate.name === elements[index]);
            if (!child) {
                return undefined;
            }
            result = child;
            ++index;
        }
        return result;
    }
    indexOfChild(child) {
        return this.children.findIndex((value, index) => value === child);
    }
    nextSiblingOf(child) {
        const index = this.indexOfChild(child);
        if (index === -1 || index >= this.children.length - 1) {
            return undefined;
        }
        return this.children[index + 1];
    }
    previousSiblingOf(child) {
        const index = this.indexOfChild(child);
        if (index < 1) {
            return undefined;
        }
        return this.children[index - 1];
    }
    nextOf(child) {
        if (!(child.parent instanceof ScopedSymbol)) {
            return undefined;
        }
        if (child.parent !== this) {
            return child.parent.nextOf(child);
        }
        if (child instanceof ScopedSymbol && child.children.length > 0) {
            return child.children[0];
        }
        const sibling = this.nextSiblingOf(child);
        if (sibling) {
            return sibling;
        }
        return this.parent.nextOf(this);
    }
}
exports.ScopedSymbol = ScopedSymbol;
class NamespaceSymbol extends ScopedSymbol {
}
exports.NamespaceSymbol = NamespaceSymbol;
class BlockSymbol extends ScopedSymbol {
}
exports.BlockSymbol = BlockSymbol;
class VariableSymbol extends TypedSymbol {
    constructor(name, value, type) {
        super(name, type);
        this.value = value;
    }
}
exports.VariableSymbol = VariableSymbol;
class LiteralSymbol extends TypedSymbol {
    constructor(name, value, type) {
        super(name, type);
        this.value = value;
    }
}
exports.LiteralSymbol = LiteralSymbol;
class ParameterSymbol extends VariableSymbol {
}
exports.ParameterSymbol = ParameterSymbol;
class RoutineSymbol extends ScopedSymbol {
    constructor(name, returnType) {
        super(name);
        this.returnType = returnType;
    }
    getVariables(localOnly = true) {
        return this.getSymbolsOfType(VariableSymbol);
    }
    getParameters(localOnly = true) {
        return this.getSymbolsOfType(ParameterSymbol);
    }
}
exports.RoutineSymbol = RoutineSymbol;
var MethodFlags;
(function (MethodFlags) {
    MethodFlags[MethodFlags["None"] = 0] = "None";
    MethodFlags[MethodFlags["Virtual"] = 1] = "Virtual";
    MethodFlags[MethodFlags["Const"] = 2] = "Const";
    MethodFlags[MethodFlags["Overwritten"] = 4] = "Overwritten";
    MethodFlags[MethodFlags["SetterOrGetter"] = 8] = "SetterOrGetter";
    MethodFlags[MethodFlags["Explicit"] = 16] = "Explicit";
})(MethodFlags = exports.MethodFlags || (exports.MethodFlags = {}));
class MethodSymbol extends RoutineSymbol {
    constructor() {
        super(...arguments);
        this.methodFlags = MethodFlags.None;
    }
}
exports.MethodSymbol = MethodSymbol;
class FieldSymbol extends VariableSymbol {
}
exports.FieldSymbol = FieldSymbol;
class ClassSymbol extends ScopedSymbol {
    constructor(name, ext, impl) {
        super(name);
        this.isStruct = false;
        this.reference = ReferenceKind.Irrelevant;
        this.extends = ext;
        this.implements = impl;
    }
    get baseTypes() { return this.extends; }
    get kind() { return TypeKind.Class; }
    getMethods(includeInherited = false) {
        return this.getSymbolsOfType(MethodSymbol);
    }
    getFields(includeInherited = false) {
        return this.getSymbolsOfType(FieldSymbol);
    }
}
exports.ClassSymbol = ClassSymbol;
class InterfaceSymbol extends ScopedSymbol {
    constructor(name, ext) {
        super(name);
        this.reference = ReferenceKind.Irrelevant;
        this.extends = ext;
    }
    get baseTypes() { return this.extends; }
    get kind() { return TypeKind.Interface; }
    getMethods(includeInherited = false) {
        return this.getSymbolsOfType(MethodSymbol);
    }
    getFields(includeInherited = false) {
        return this.getSymbolsOfType(FieldSymbol);
    }
}
exports.InterfaceSymbol = InterfaceSymbol;
class ArrayType extends Symbol {
    constructor(name, referenceKind, elemType, size = 0) {
        super(name);
        this.referenceKind = referenceKind;
        this.elementType = elemType;
        this.size = size;
    }
    get baseTypes() { return []; }
    get kind() { return TypeKind.Array; }
    get reference() { return this.referenceKind; }
}
exports.ArrayType = ArrayType;
class SymbolTable extends ScopedSymbol {
    constructor(name, options) {
        super(name);
        this.options = options;
        this.dependencies = new Set();
    }
    get info() {
        return {
            dependencyCount: this.dependencies.size,
            symbolCount: this.children.length,
        };
    }
    clear() {
        super.clear();
        this.dependencies.clear();
    }
    addDependencies(...tables) {
        tables.forEach((value, key) => {
            this.dependencies.add(value);
        });
    }
    removeDependency(table) {
        if (this.dependencies.has(table)) {
            this.dependencies.delete(table);
        }
    }
    addNewSymbolOfType(t, parent, ...args) {
        const result = new t(...args);
        if (!parent || parent === this) {
            this.addSymbol(result);
        }
        else {
            parent.addSymbol(result);
        }
        return result;
    }
    addNewNamespaceFromPath(parent, path, delimiter = ".") {
        return __awaiter(this, void 0, void 0, function* () {
            const parts = path.split(delimiter);
            let i = 0;
            let currentParent = (parent === undefined) ? this : parent;
            while (i < parts.length - 1) {
                let namespace = yield currentParent.resolve(parts[i], true);
                if (namespace === undefined) {
                    namespace = this.addNewSymbolOfType(NamespaceSymbol, currentParent, parts[i]);
                }
                currentParent = namespace;
                ++i;
            }
            return this.addNewSymbolOfType(NamespaceSymbol, currentParent, parts[parts.length - 1]);
        });
    }
    addNewNamespaceFromPathSync(parent, path, delimiter = ".") {
        const parts = path.split(delimiter);
        let i = 0;
        let currentParent = (parent === undefined) ? this : parent;
        while (i < parts.length - 1) {
            let namespace = currentParent.resolveSync(parts[i], true);
            if (namespace === undefined) {
                namespace = this.addNewSymbolOfType(NamespaceSymbol, currentParent, parts[i]);
            }
            currentParent = namespace;
            ++i;
        }
        return this.addNewSymbolOfType(NamespaceSymbol, currentParent, parts[parts.length - 1]);
    }
    getAllSymbols(t, localOnly = false) {
        const _super = Object.create(null, {
            getAllSymbols: { get: () => super.getAllSymbols }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield _super.getAllSymbols.call(this, t, localOnly);
            if (!localOnly) {
                const dependencyResults = yield Promise.all([...this.dependencies].map((dependency) => (dependency.getAllSymbols(t, localOnly))));
                dependencyResults.forEach((value) => {
                    result.push(...value);
                });
            }
            return result;
        });
    }
    getAllSymbolsSync(t, localOnly = false) {
        const result = super.getAllSymbolsSync(t, localOnly);
        if (!localOnly) {
            this.dependencies.forEach((dependency) => {
                result.push(...dependency.getAllSymbolsSync(t, localOnly));
            });
        }
        return result;
    }
    symbolWithContext(context) {
        return __awaiter(this, void 0, void 0, function* () {
            const findRecursive = (symbol) => {
                if (symbol.context === context) {
                    return symbol;
                }
                if (symbol instanceof ScopedSymbol) {
                    for (const child of symbol.children) {
                        const result = findRecursive(child);
                        if (result) {
                            return result;
                        }
                    }
                }
                return undefined;
            };
            let symbols = yield this.getAllSymbols(Symbol);
            for (const symbol of symbols) {
                const result = findRecursive(symbol);
                if (result) {
                    return result;
                }
            }
            for (const dependency of this.dependencies) {
                symbols = yield dependency.getAllSymbols(Symbol);
                for (const symbol of symbols) {
                    const result = findRecursive(symbol);
                    if (result) {
                        return result;
                    }
                }
            }
            return undefined;
        });
    }
    symbolWithContextSync(context) {
        const findRecursive = (symbol) => {
            if (symbol.context === context) {
                return symbol;
            }
            if (symbol instanceof ScopedSymbol) {
                for (const child of symbol.children) {
                    const result = findRecursive(child);
                    if (result) {
                        return result;
                    }
                }
            }
            return undefined;
        };
        let symbols = this.getAllSymbolsSync(Symbol);
        for (const symbol of symbols) {
            const result = findRecursive(symbol);
            if (result) {
                return result;
            }
        }
        for (const dependency of this.dependencies) {
            symbols = dependency.getAllSymbolsSync(Symbol);
            for (const symbol of symbols) {
                const result = findRecursive(symbol);
                if (result) {
                    return result;
                }
            }
        }
        return undefined;
    }
    resolve(name, localOnly = false) {
        const _super = Object.create(null, {
            resolve: { get: () => super.resolve }
        });
        return __awaiter(this, void 0, void 0, function* () {
            let result = yield _super.resolve.call(this, name, localOnly);
            if (!result && !localOnly) {
                for (const dependency of this.dependencies) {
                    result = yield dependency.resolve(name, false);
                    if (result) {
                        return result;
                    }
                }
            }
            return result;
        });
    }
    resolveSync(name, localOnly = false) {
        let result = super.resolveSync(name, localOnly);
        if (!result && !localOnly) {
            for (const dependency of this.dependencies) {
                result = dependency.resolveSync(name, false);
                if (result) {
                    return result;
                }
            }
        }
        return result;
    }
}
exports.SymbolTable = SymbolTable;
